### PR TITLE
Hcall and Qcall implementation update

### PR DESCRIPTION
**Pull request by 123abcpp**
Use shared memory for message exchange, ensure all the data in the message and the the data pointed by the pointer in the message are allocated in the shared memory.

### CODE REVIEW

File: qlib/kernel/Kernel.rs [hunk 1]
```diff
@@ -39,44 +39,78 @@ impl HostSpace {
...
-        HyperCall64_init(HYPERCALL_IOWAIT, 0, 0, 0, 0);
+        HyperCall64(HYPERCALL_IOWAIT, 0, 0, 0, 0);
     }
 
     pub fn Hlt() {
         HyperCall64(HYPERCALL_HLT, 0, 0, 0, 0);
     }
-
+    
+    /*
```
Comment (chl337) [inline qlib/kernel/Kernel.rs]: If it's not needed, please remove it.

File: qkernel/src/kernel_def.rs [hunk 2]
```diff
@@ -296,20 +296,35 @@ impl HostSpace {
...
-            msg: msg,
-        };
-
-        let addr = &qMsg as *const _ as u64;
-        let om = HostOutputMsg::QCall(addr);
-
-        super::SHARESPACE.AQCall(&om);
+        let qmsg_size = core::mem::size_of::<QMsg>();
+        let msg_size = core::mem::size_of::<Msg>();
+        let qmsg_ptr = unsafe { GLOBAL_ALLOCATOR.AllocSharedBuf(qmsg_size,0x80) as *mut QMsg };
```
Comment (chl337) [inline qkernel/src/kernel_def.rs]: I suggest defining the hard-coded _0x80_ in a variable.

File: qkernel/src/kernel_def.rs [hunk 3]
```diff
@@ -328,12 +343,12 @@ impl HostSpace {
         event.globalLock = lock;
         event.ret = 0;
         event.msg = new_msg;
-
+
```
Comment (chl337) [inline qkernel/src/kernel_def.rs:346]: Seems like an unintentional tab.

File: qlib/kernel/Kernel.rs [hunk 4]
```diff
@@ -46,79 +46,67 @@ impl HostSpace {
         HyperCall64(HYPERCALL_HLT, 0, 0, 0, 0);
     }
     
-    /* 
+    //Here exists memory leak, vec in shared struct cannot be dealloc
```
Comment (chl337) [inline qlib/kernel/Kernel.rs:53]: This is unclear to me. Is the leak caused because of this function or because _of DeallocShareBuf_?

File: qkernel/src/kernel_def.rs [hunk 5]
```diff
@@ -208,24 +208,77 @@ pub fn Invlpg(addr: u64) {
...
+#[derive(Debug, Copy, Clone, Default)]
+#[repr(C)]
+pub struct SharePara {
+    pub para1: u64,
+    pub para2: u64,
+    pub para3: u64,
+    pub para4: u64,
+}
+
+pub const SHAREPARA_COUNT: u64 = MemoryDef::PAGE_SIZE/core::mem::size_of::<SharePara>() as u64;
```
Comment (chl337) [inline qkernel/src/kernel_def.rs]: This seems strange. Shouldn't it be four?

File: qlib/kernel/Kernel.rs [hunk 6]
```diff
@@ -823,8 +844,10 @@ impl HostSpace {
     }
 
     pub fn VcpuWait() -> i64 {
-        let mut ret: i64 = 0;
-        HyperCall64(HYPERCALL_VCPU_WAIT, 0, 0, &mut ret as *mut _ as u64, 0);
+        let ret_ptr =  unsafe{ GLOBAL_ALLOCATOR.AllocSharedBuf(size_of::<i64>(),0x80) as *mut i64};
+        (HYPERCALL_VCPU_WAIT, 0, 0, ret_ptr as u64, 0);
```
Comment (chl337) [inline qlib/kernel/Kernel.rs]: Are you missing a function name here?

**Comment by chl337** [comment]
Can you please fix the indentation, e.g., on _unsafe{}_ blocks?

**Comment by shrik3** [comment]
> Can you please fix the indentation, e.g., on _unsafe{}_ blocks?

I'd suggest we add a rustfmt specs for this codebase for consistent (and most importantly, automated) coding style.

File: qlib/kernel/Kernel.rs (hunk 4 above)
Comment (123abcpp) [reply qlib/kernel/Kernel.rs]: The process struct contains vec, initialized in qvisor by pointer in unsafe, thus no one manages these vecs, I only dealloc the process struct but not the addresses pointed by the pointers inside.

File: qkernel/src/kernel_def.rs (hunk 5 above)
Comment (123abcpp) [reply qkernel/src/kernel_def.rs]: I leave a page for shared parameters, the count is the maximum number of SharePara structs in this page.

File: qlib/kernel/Kernel.rs (hunk 6 above)
Comment (123abcpp) [reply qlib/kernel/Kernel.rs]: > Are you missing a function name here?

Yeah, already fixed

**Comment by 123abcpp** [comment]
> Can you please fix the indentation, e.g., on _unsafe{}_ blocks?
Could you give me a more detailed example? How would you like me to fix these unsafe blocks for indentation?

File: qkernel/src/kernel_def.rs [hunk 7]
```diff
@@ -208,24 +209,54 @@ pub fn Invlpg(addr: u64) {
...
-            in("rsi") para1,
-            in("rcx") para2,
-            in("rdi") para3,
-            in("r10") para4
         )
     }
 }
 
+//VCPU 0 did not set gs as corresponding Vcpulocal address, thus use another function to use default position
+#[cfg(target_arch = "x86_64")]
```
Comment (yaoxin1995) [inline qkernel/src/kernel_def.rs:238]: why don't you setup the VCPU0's number after vcpu0 enters the guest? In this case you only need one `HyperCall64` function

File: qkernel/src/lib.rs [hunk 8]
```diff
@@ -480,27 +479,29 @@ pub extern "C" fn rust_main(
...
 
         SHARESPACE.SetValue(shared_space as u64);
         SingletonInit();
-
+        //HyperCall64 can be called after here, since gs set in SingletonInit
         SetVCPCount(vcpuCnt as usize);
         VCPU_ALLOCATOR.Print();
         VCPU_ALLOCATOR.Initializated();
 
+        info!("123");
```
Comment (yaoxin1995) [inline qkernel/src/lib.rs:504]: remove the leftover

File: qlib/kernel/Kernel_cc.rs [hunk 9]
```diff
@@ -0,0 +1,2085 @@
...
+            Msg::LoadProcessKernel(LoadProcessKernel {
+                processAddr: process_ptr as u64,
+            }),
+            GUEST_HOST_SHARED_ALLOCATOR,
+        );
+        let ret = HostSpace::Call(&mut msg, false) as i64;
+        let private_process = unsafe { &mut *(processAddr as *mut Process) };
+        private_process.clone_from_shared(process_ptr);
+        //info!("###Cloned Process:{:#?}\n",private_process);
+        //Here could have memory leak since
```
Comment (yaoxin1995) [inline qlib/kernel/Kernel_cc.rs:56]: Is there still a memory leak? Shouldn't the shared memory be dallocated in Global_Allocator when the fuction drops new_process?

File: qkernel/makefile [hunk 10]
```diff
@@ -24,12 +32,27 @@ $(kernel_debug): kernel_debug $(assembly_object_files)
...
 kernel:
 	CARGO_TARGET_DIR=../target cargo +nightly-2022-08-11-x86_64-unknown-linux-gnu xbuild --target $(arch)-qkernel.json --release
 
 kernel_debug:
 	CARGO_TARGET_DIR=../target cargo +nightly-2022-08-11-x86_64-unknown-linux-gnu xbuild --target $(arch)-qkernel.json
 
+cc_kernel:
+	CARGO_TARGET_DIR=../target cargo +nightly-2022-08-11-x86_64-unknown-linux-gnu xbuild --target $(arch)-qkernel.json --release --features cc
+
+cc_kernel_debug:
```
Comment (shrik3) [inline qkernel/makefile:53]: So we are adding two series of targets, all these complexities only too pass the "features += cc" to the cargo build arguments?

I thought it was possible to do that via environment variables but it turns not out to be the case. [This feature request](https://github.com/rust-lang/cargo/issues/4829) has been there for ~7 years, shame.

Comment (shrik3) [reply qkernel/makefile:53]: > not good idea, because containerd is not capable to pass extral enviroment variable for cc

You are misunderstanding me. It's only about setting compile-time config. For example to specify a build target you could do `RUSTFLAGS="-C target-cpu=neoverse-n1" cargo build`, this is surprisingly not the case for the `--features` argument.

File: qkernel/src/kernel_def.rs (hunk 7 above)
Comment (123abcpp) [reply qkernel/src/kernel_def.rs:258]: > why don't you setup the VCPU0's number after vcpu0 enters the guest? In this case you only need one `HyperCall64` function

Swapgs to set the vcpu number rely on the sharedspace, we still need a hypercall to initialize the sharedspace struct.

File: qlib/kernel/Kernel_cc.rs (hunk 9 above)
Comment (123abcpp) [reply qlib/kernel/Kernel_cc.rs:56]: Yes, the memory leak should not exist here now.

File: qlib/kernel/quring/uring_async.rs [hunk 11]
```diff
@@ -1060,7 +1061,11 @@ impl AIOWrite {
         eventfops: Option<EventOperations>,
     ) -> Result<Self> {
         let vec = task.CopyInVec(cb.buf, cb.bytes as usize)?;
-        let buf = DataBuff { buf: vec };
+        let mut shared_vec = Vec::new_in(GUEST_HOST_SHARED_ALLOCATOR);
```
Comment (yaoxin1995) [inline qlib/kernel/quring/uring_async.rs:1064]: Also changed ucall?

File: qlib/linux_def.rs [hunk 12]
```diff
@@ -2459,7 +2461,7 @@ pub struct IoVec {
 
 #[derive(Debug)]
 pub struct DataBuff {
-    pub buf: Vec<u8>,
+    pub buf: Vec<u8,GuestHostSharedAllocator>,
```
Comment (yaoxin1995) [inline qlib/linux_def.rs:2464]: Why the change?

File: qlib/kernel/boot/loader.rs [hunk 13]
```diff
@@ -251,7 +251,7 @@ impl Loader {
             &procArgs.Filename,
             &paths,
         )?;
-
+        info!("##Load 1");
```
[the remark this answers is not in the document]
Comment (123abcpp) [reply qlib/kernel/boot/loader.rs:254]: Removed

File: qlib/kernel/quring/uring_async.rs (hunk 11 above)
Comment (123abcpp) [reply qlib/kernel/quring/uring_async.rs:1064]: I did not remember, maybe I changed this when I modified DataBuff struct here.

File: qlib/linux_def.rs (hunk 12 above)
Comment (123abcpp) [reply qlib/linux_def.rs:2464]: DataBuff is used for IO.

File: qlib/kernel/Kernel_cc.rs [hunk 14]
```diff
@@ -0,0 +1,2083 @@
...
+use crate::GLOBAL_ALLOCATOR;
+use crate::GUEST_HOST_SHARED_ALLOCATOR;
+
+extern "C" {
+    pub fn rdtsc() -> i64;
+}
+
+impl HostSpace {
+    pub fn LoadProcessKernel_cc(processAddr: u64) -> i64 {
+        let mut new_process = Box::new_in(Process::default(), GUEST_HOST_SHARED_ALLOCATOR);
```
[the remark this answers is not in the document]
Comment (123abcpp) [reply qlib/kernel/Kernel_cc.rs:43]: @yaoxin1995 A new vec will not apply for memory after data is pushed in, and the Process::default() is a struct with no data inside derived by the default trait. When returning to qvisor, it will resize the vecs inside when pushing data, using its global allocator, which is the shared one.
